cni/pkg/install-cni/cmd/root.go

```diff
@@ -0,0 +1,131 @@
+	"istio.io/istio/cni/pkg/install-cni/pkg/install"
+	"istio.io/pkg/log"
+)
+
+var rootCmd = &cobra.Command{
+	Use:   "install-cni",
+	Short: "Install and configure CNI on a node",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		cfg, err := constructConfig()
+		if err != nil {
+			return err
+		}
+		return install.Run(cfg)
+	},
+}
+
+// GetCommand returns the main cobra.Command object for this application
+func GetCommand() *cobra.Command {
+	return rootCmd
+}
+
+func init() {
+	viper.AutomaticEnv()
+	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+
+	registerStringParameter(constants.CNINetDir, "/etc/cni/net.d", "Directory on the host where CNI networks are installed")
+	registerStringParameter(constants.CNIConfName, "", "Name of the CNI configuration file")
+	registerBooleanParameter(constants.ChainedCNIPlugin, true, "Whether to install CNI plugin as a chained or standalone")
+	registerStringParameter(constants.CNINetworkConfig, "", "CNI config template as a string")
+
+	// Not configurable in CNI helm charts
+	registerStringParameter(constants.MountedCNINetDir, "/host/etc/cni/net.d", "Directory on the container where CNI networks are installed")
+	registerBooleanParameter(constants.Sleep, true,
+		"Whether to sleep and wait for modifications after installation to prevent Kubernetes from restarting the pod repeatedly")
+	registerStringParameter(constants.CNINetworkConfigFile, "", "CNI config template as a file")
+	registerStringParameter(constants.LogLevel, "warn", "Fallback value for log level in CNI config file, if not specified in helm template")
+	registerStringParameter(constants.KubecfgFilename, "ZZZ-istio-cni-kubeconfig", "Name of the kubeconfig file")
+	registerIntegerParameter(constants.KubeconfigMode, constants.DefaultKubeconfigMode, "File mode of the kubeconfig file")
+	registerStringParameter(constants.KubeCAFile, "", "CA file for kubeconfig. Defaults to the pod one")
+	registerBooleanParameter(constants.SkipTLSVerify, false, "Whether to use insecure TLS in kubeconfig file")
+	registerBooleanParameter(constants.UpdateCNIBinaries, true, "Update binaries")
+	registerStringArrayParameter(constants.SkipCNIBinaries, []string{}, "Binaries that should not be installed")
+}
+
+func registerStringParameter(name, value, usage string) {
+	rootCmd.Flags().String(name, value, usage)
+	bindViper(name)
+}
+
+func registerStringArrayParameter(name string, value []string, usage string) {
+	rootCmd.Flags().StringArray(name, value, usage)
+	bindViper(name)
+}
+
+func registerIntegerParameter(name string, value int, usage string) {
+	rootCmd.Flags().Int(name, value, usage)
+	bindViper(name)
+}
+
+func registerBooleanParameter(name string, value bool, usage string) {
+	rootCmd.Flags().Bool(name, value, usage)
+	bindViper(name)
+}
+
+func bindViper(name string) {
+	if err := viper.BindPFlag(name, rootCmd.Flags().Lookup(name)); err != nil {
+		log.Errora(err)
+		os.Exit(1)
+	}
+}
+
+func constructConfig() (*config.Config, error) {
```

Comment:
Move `constructConfig` to `pkg/config/config.go`.

Reply:
and probably call it `New`

Reply:
I disagree about moving this function. This is the right place for it. It's strongly coupled with the rest of this file (set and names of environment variables) and they must always change together. So they should remain together in the same file.

Reply:
Keeping the `structs` and the `instance` of it together is a good practice already followed by lot of Kubernetes libs eg: [client-go](https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/api/types.go#L251).

However, I see that almost all of the values are either fetched from  `constants` or from `env`. So I believe there is no harm in moving to `pkg/config/config.go` and naming it `NewConfig` ?

Reply:
I agree with @rlenglet since the purpose of the function is to create an instance of Config based on the environment which the root command was called. This choice is also used in `istio-iptables` with a [`constructConfig()` function](https://github.com/istio/istio/blob/7e486c75d142b7ce16e09fd830850a7502e05d74/tools/istio-iptables/pkg/cmd/root.go#L83).